osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs

```diff
@@ -111,50 +115,160 @@ public abstract class DrawableHitObject : SkinReloadableDrawable
         }
 
         protected override void LoadAsyncComplete()
         {
             base.LoadAsyncComplete();
 
-            LoadSamples();
+            if (HitObject != null)
+                Apply(HitObject);
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            StartTimeBindable.BindValueChanged(_ => updateState(State.Value, true));
+            comboIndexBindable.BindValueChanged(_ => updateComboColour(), true);
+
+            updateState(ArmedState.Idle, true);
+        }
+
+        /// <summary>
+        /// Applies a new <see cref="HitObject"/> to be represented by this <see cref="DrawableHitObject"/>.
+        /// </summary>
+        /// <param name="hitObject">The <see cref="HitObject"/> to apply.</param>
+        public void Apply(HitObject hitObject)
+        {
+            free();
+
+            HitObject = hitObject ?? throw new InvalidOperationException($"Cannot apply a null {nameof(HitObject)}.");
+
+            // Copy any existing result from the hitobject (required for rewind / judgement revert).
+            Result = HitObject.Result;
+
+            // Ensure this DHO has a result.
+            Result ??= CreateResult(HitObject.CreateJudgement())
+                       ?? throw new InvalidOperationException($"{GetType().ReadableName()} must provide a {nameof(JudgementResult)} through {nameof(CreateResult)}.");
+
+            // Ensure the hitobject has a result.
+            HitObject.Result = Result;
+
+            foreach (var h in HitObject.NestedHitObjects)
+            {
+                var drawableNested = CreateNestedHitObject(h) ?? throw new InvalidOperationException($"{nameof(CreateNestedHitObject)} returned null for {h.GetType().ReadableName()}.");
+
+                drawableNested.OnNewResult += (d, r) => OnNewResult?.Invoke(d, r);
+                drawableNested.OnRevertResult += (d, r) => OnRevertResult?.Invoke(d, r);
+                drawableNested.ApplyCustomUpdateState += (d, j) => ApplyCustomUpdateState?.Invoke(d, j);
+
+                nestedHitObjects.Value.Add(drawableNested);
+                AddNestedHitObject(drawableNested);
+            }
+
+            StartTimeBindable.BindTo(HitObject.StartTimeBindable);
+            if (HitObject is IHasComboInformation combo)
+                comboIndexBindable.BindTo(combo.ComboIndexBindable);
+
+            samplesBindable.BindTo(HitObject.SamplesBindable);
+            samplesBindable.BindCollectionChanged(onSamplesChanged, true);
 
             HitObject.DefaultsApplied += onDefaultsApplied;
 
-            startTimeBindable = HitObject.StartTimeBindable.GetBoundCopy();
-            startTimeBindable.BindValueChanged(_ => updateState(State.Value, true));
+            // If not loaded, the state update happens in LoadComplete(). Otherwise, the update is scheduled to allow for lifetime updates.
+            if (IsLoaded)
+                Schedule(() => updateState(ArmedState.Idle, true));
```

Comment:
Maybe it should be ensured that this doesn't get called more than once? Not sure how realistic but I can see a scenario wherein this could get triggered multiple times but I don't think it ever really needs to.

Reply:
Would that ever be a problem? The rest of the code here is still going to run 1000s of times if it was called that many times within a single frame, of which sample reconstruction is by far the most expensive.

We decided a while back with the editor to keep the amount of debouncing to a minimum, and I don't think it's worth adding it here.

Reply:
Fair point, was just asking. Don't recall being aware of such a decision being made.